Sources/Maintenance.php

```diff
@@ -0,0 +1,836 @@
+	{
+		Security::frameOptionsHeader('SAMEORIGIN');
+		self::$theme_dir = dirname(SMF_SETTINGS_FILE) . '/Themes/default';
+
+		if (!defined('DISABLE_TOOL_SECURITY')) {
+			define('DISABLE_TOOL_SECURITY', false);
+		}
+		self::$disable_security = DISABLE_TOOL_SECURITY;
+
+		// This might be overwritten by the tool, but we need a default value.
+		self::$context['started'] = (int) TIME_START;
+	}
+
+	/**
+	 * This is the main call to get stuff done.
+	 *
+	 * @var int The tool type we are running.
+	 */
+	public function execute(int $type): void
+	{
+		if (!self::toolIsValid($type)) {
+			die('Invalid Tool selected');
+		}
+
+		// Handle the CLI.
+		if (Sapi::isCLI()) {
+			self::parseCliArguments();
+		}
+
+		/** @var \SMF\Maintenance\ToolsInterface&\SMF\Maintenance\ToolsBase $tool_class */
+		$tool_class = '\\SMF\\Maintenance\\Tools\\' . self::$valid_tools[$type];
+
+		require_once Config::$sourcedir . '/Maintenance/Tools/' . self::$valid_tools[$type] . '.php';
+		self::$tool = new $tool_class();
+
+		/** @var \SMF\Maintenance\TemplateInterface $template_class */
+		$template_class = '\\SMF\\Maintenance\\Template\\' . self::$valid_tools[$type];
+
+		require_once Config::$sourcedir . '/Maintenance/Template/' . self::$valid_tools[$type] . '.php';
+		self::$template = new $template_class();
+
+		// This is really quite simple; if ?delete is on the URL, delete the installer...
+		if (isset($_GET['delete'])) {
+			self::$tool->deleteTool();
+
+			exit;
+		}
+
+		/** @var \SMF\Maintenance\Step $step */
+		foreach (self::$tool->getSteps() as $num => $step) {
+			if ($num >= self::getCurrentStep()) {
+				// The current weight of this step in terms of overall progress.
+				self::$context['step_weight'] = $step->getProgress();
+
+				// Make sure we reset the skip button.
+				self::$context['skip'] = false;
+
+				// Call the step and if it returns false that means pause!
+				if (
+					method_exists(self::$tool, $step->getFunction())
+					&& self::$tool->{$step->getFunction()}() === false
+				) {
+					break;
+				}
+
+				// Time to move on.
+				self::setCurrentStep();
+				self::setCurrentSubStep(0);
+				self::setCurrentStart(0);
+
+				// No warnings pass on.
+				self::$context['warning'] = '';
+			}
+
+			self::$overall_percent += (int) $step->getProgress();
+		}
+
+		// Last chance to set our template.
+		if (self::$sub_template === '') {
+			self::$sub_template = self::$tool->getSteps()[self::getCurrentStep()]->getFunction();
+		}
+
+		// Make a final call before we are done..
+		self::$tool->preExit();
+
+		self::exit();
+	}
+
+	/***********************
+	 * Public static methods
+	 ***********************/
+
+	/**
+	 * See if we think they have already installed it?
+	 *
+	 * @return bool Whether we believe SMF has been installed.
+	 */
+	public static function isInstalled(): bool
+	{
+		foreach (['image_proxy_secret', 'db_passwd', 'boardurl'] as $var) {
+			if (Config::${$var} === Config::$settings_defs[$var]['default']) {
```

Comment:
@Sesquipedalian $settings_defs is protected and is not accessible.

Reply:
Oh, right.